### PR TITLE
Restructure components, and add Azure hosting bits

### LINKED ISSUE #311: 'IDistributedApplicationBuilder' does not contain a definition for 'AddAzureCosmosDB
### Type of issue

Code doesn't work

### Description

Following instructions [here](https://learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-entity-framework-component?tabs=dotnet-cli#orchestration) for orchestration I get the error:

```sh
error CS1061: 'IDistributedApplicationBuilder' does not contain a definition for 'AddAzureCosmosDB' and no accessible extension method 'AddAzureCosmosDB' accepting a first argument of type 'IDistributedApplicationBuilder' could be found (are you missing a using directive or an assembly reference?) 
```

I have the following packages installed
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <IsAspireHost>true</IsAspireHost>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Aspire.Hosting" Version="8.0.0-preview.2.23619.3" />
    <PackageReference Include="Aspire.Microsoft.EntityFrameworkCore.Cosmos" Version="8.0.0-preview.2.23619.3" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\TodoApi\TodoApi.csproj" />
  </ItemGroup>

</Project>
```

and my AppHost looks like:

```c#
using Aspire.Microsoft.EntityFrameworkCore.Cosmos;

var builder = DistributedApplication.CreateBuilder(args);

var cosmosdbService = builder
    .AddAzureCosmosDB("cosmosdb")
    .AddDatabase("todos");
                            
builder.AddProject<Projects.TodoApi>("todoapi")
    .WithReference(cosmosdbService); ;

builder.Build().Run();
```

### Page URL

https://learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-entity-framework-component?tabs=dotnet-cli

### Content source URL

https://github.com/dotnet/docs-aspire/blob/main/docs/database/azure-cosmos-db-entity-framework-component.md

### Document Version Independent Id

8b585800-8c7b-df72-3125-2776e1b1375e

### Article author

@IEvangelist

---
[Associated WorkItem - 202718](https://dev.azure.com/msft-skilling/Content/_workitems/edit/202718)

### LINKED ISSUE #314: Restructure component articles
We need to disambiguate the App Host APIs and code examples from the consuming library APIs. We should also, restructure the component articles such that their example usages are immediately followed by the App host section, with config section following that.

---
[Associated WorkItem - 202717](https://dev.azure.com/msft-skilling/Content/_workitems/edit/202717)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Restructure components, example usage, then app host, then config.
- Add `Aspire.Hosting.Azure` for Azure components.
- Normalize app host, AppHost, and Orchestrator terminology.

Issues fixed by this PR:

- Fixes #311
- Fixes #314

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/caching/stackexchange-redis-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/caching/stackexchange-redis-component.md) | [.NET Aspire StackExchange Redis component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-component?branch=pr-en-us-315) |
| [docs/caching/stackexchange-redis-distributed-caching-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/caching/stackexchange-redis-distributed-caching-component.md) | [.NET Aspire StackExchange Redis distributed caching component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-distributed-caching-component?branch=pr-en-us-315) |
| [docs/caching/stackexchange-redis-output-caching-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/caching/stackexchange-redis-output-caching-component.md) | [.NET Aspire StackExchange Redis output caching component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-output-caching-component?branch=pr-en-us-315) |
| [docs/database/azure-cosmos-db-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/azure-cosmos-db-component.md) | [docs/database/azure-cosmos-db-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-component?branch=pr-en-us-315) |
| [docs/database/azure-cosmos-db-entity-framework-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/azure-cosmos-db-entity-framework-component.md) | [.NET Aspire Microsoft Entity Framework Core Cosmos DB component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-entity-framework-component?branch=pr-en-us-315) |
| [docs/database/mongodb-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/mongodb-component.md) | [.NET Aspire MongoDB database component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/mongodb-component?branch=pr-en-us-315) |
| [docs/database/mysql-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/mysql-component.md) | [.NET Aspire MySQL database component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/mysql-component?branch=pr-en-us-315) |
| [docs/database/postgresql-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/postgresql-component.md) | [docs/database/postgresql-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/postgresql-component?branch=pr-en-us-315) |
| [docs/database/postgresql-entity-framework-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/postgresql-entity-framework-component.md) | [.NET Aspire PostgreSQL Entity Framework Core component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/postgresql-entity-framework-component?branch=pr-en-us-315) |
| [docs/database/sql-server-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/sql-server-component.md) | [.NET Aspire SQL Server component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/sql-server-component?branch=pr-en-us-315) |
| [docs/database/sql-server-components.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/sql-server-components.md) | [Tutorial: Connect an ASP.NET Core app to SQL Server using .NET Aspire and Entity Framework Core](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/sql-server-components?branch=pr-en-us-315) |
| [docs/database/sql-server-entity-framework-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/database/sql-server-entity-framework-component.md) | [docs/database/sql-server-entity-framework-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/sql-server-entity-framework-component?branch=pr-en-us-315) |
| [docs/fundamentals/components-overview.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/fundamentals/components-overview.md) | [.NET Aspire components overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/fundamentals/components-overview?branch=pr-en-us-315) |
| [docs/get-started/aspire-overview.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/get-started/aspire-overview.md) | [.NET Aspire overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview?branch=pr-en-us-315) |
| [docs/messaging/azure-service-bus-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/messaging/azure-service-bus-component.md) | [.NET Aspire Azure Service Bus component](https://review.learn.microsoft.com/en-us/dotnet/aspire/messaging/azure-service-bus-component?branch=pr-en-us-315) |
| [docs/messaging/messaging-components.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/messaging/messaging-components.md) | [[Passwordless (Recommended)](#tab/passwordless)](https://review.learn.microsoft.com/en-us/dotnet/aspire/messaging/messaging-components?branch=pr-en-us-315) |
| [docs/messaging/rabbitmq-client-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/messaging/rabbitmq-client-component.md) | [docs/messaging/rabbitmq-client-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/messaging/rabbitmq-client-component?branch=pr-en-us-315) |
| [docs/security/azure-security-key-vault-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/security/azure-security-key-vault-component.md) | [.NET Aspire Azure Key Vault component](https://review.learn.microsoft.com/en-us/dotnet/aspire/security/azure-security-key-vault-component?branch=pr-en-us-315) |
| [docs/storage/azure-storage-blobs-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/storage/azure-storage-blobs-component.md) | [.NET Aspire Azure Blob Storage component](https://review.learn.microsoft.com/en-us/dotnet/aspire/storage/azure-storage-blobs-component?branch=pr-en-us-315) |
| [docs/storage/azure-storage-components.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/storage/azure-storage-components.md) | [Tutorial: Connect an ASP.NET Core app to .NET Aspire storage components](https://review.learn.microsoft.com/en-us/dotnet/aspire/storage/azure-storage-components?branch=pr-en-us-315) |
| [docs/storage/azure-storage-queues-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/storage/azure-storage-queues-component.md) | [.NET Aspire Azure Queue Storage component](https://review.learn.microsoft.com/en-us/dotnet/aspire/storage/azure-storage-queues-component?branch=pr-en-us-315) |
| [docs/storage/azure-storage-tables-component.md](https://github.com/dotnet/docs-aspire/blob/ec45ab4dc0ae6d2945151768c9a15fad6d44edbe/docs/storage/azure-storage-tables-component.md) | [.NET Aspire Azure Data Tables component](https://review.learn.microsoft.com/en-us/dotnet/aspire/storage/azure-storage-tables-component?branch=pr-en-us-315) |

</details>

### CODE REVIEW

File: docs/caching/stackexchange-redis-component.md [hunk 1]
```diff
@@ -32,7 +32,7 @@ For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-pac
 
 ## Example usage
 
-In the _Program.cs_ file of your project, call the <xref:Aspire.Hosting.RedisBuilderExtensions.AddRedis%2A> extension to register a `IConnectionMultiplexer` for use via the dependency injection container.
+In the _Program.cs_ file of your component consuming project, call the <xref:Aspire.Hosting.RedisBuilderExtensions.AddRedis%2A> extension to register a `IConnectionMultiplexer` for use via the dependency injection container.
```
Comment (gewarren) [inline docs/caching/stackexchange-redis-component.md:35]:
```suggestion
In the _Program.cs_ file of your component-consuming project, call the <xref:Aspire.Hosting.RedisBuilderExtensions.AddRedis%2A> extension to register a `IConnectionMultiplexer` for use via the dependency injection container.
```

File: docs/caching/stackexchange-redis-distributed-caching-component.md [hunk 2]
```diff
@@ -32,7 +32,7 @@ For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-pac
 
 ## Example usage
 
-In the _Program.cs_ file of your project, call the <xref:Microsoft.Extensions.Hosting.AspireRedisDistributedCacheExtensions.AddRedisDistributedCache%2A> extension to register the required services for distributed caching and add a <xref:Microsoft.Extensions.Caching.Distributed.IDistributedCache> for use via the dependency injection container.
+In the _Program.cs_ file of your component consuming project, call the <xref:Microsoft.Extensions.Hosting.AspireRedisDistributedCacheExtensions.AddRedisDistributedCache%2A> extension to register the required services for distributed caching and add a <xref:Microsoft.Extensions.Caching.Distributed.IDistributedCache> for use via the dependency injection container.
```
Comment (gewarren) [inline docs/caching/stackexchange-redis-distributed-caching-component.md:35]:
```suggestion
In the _Program.cs_ file of your component-consuming project, call the <xref:Microsoft.Extensions.Hosting.AspireRedisDistributedCacheExtensions.AddRedisDistributedCache%2A> extension to register the required services for distributed caching and add a <xref:Microsoft.Extensions.Caching.Distributed.IDistributedCache> for use via the dependency injection container.
```

**Review by gewarren — APPROVED** [review]

**Comment by bradygaster** [comment]
@IEvangelist this is great, exciting to see the updates. Could we do one more tweak on these (sorry!)? Each one should have this sort of approach:

1. Add the NuGet package to the app host project
2. Add the orchestration code to the app host project
3. Add the NuGet package the project-that-will-use-the-feature-or-component
4. Add the code to the project

In the Redis one it seemed close to this, but App Host came after target project. I think we should hit the app host first.

**Comment by IEvangelist** [comment]
> @IEvangelist this is great, exciting to see the updates. Could we do one more tweak on these (sorry!)? Each one should have this sort of approach:
> 
> 1. Add the NuGet package to the app host project
> 2. Add the orchestration code to the app host project
> 3. Add the NuGet package the project-that-will-use-the-feature-or-component
> 4. Add the code to the project
> 
> In the Redis one it seemed close to this, but App Host came after target project. I think we should hit the app host first.

@bradygaster 

I thought we decided to prioritize the component-consuming project first, since these can technically be consumed without any app host at all. The order I had in my notes was:

- component-consuming project example
- app host project
  - the only time a NuGet package is needed here is for Azure hosting, which was added in the PR
- config

**Comment by bradygaster** [comment]
@IEvangelist i'd say for any component that needs to be in the App Host before it'll work in the actual project, the App Host should come first. Apologies if I didn't clarify that in the discussion. If someone, say, adds a Redis component to their front-end web project before having the call to `AddRedis` in the App Host, things will break when they F5, right? I'm optimizing for the F5 experience.

**Comment by IEvangelist** [comment]
> @IEvangelist i'd say for any component that needs to be in the App Host before it'll work in the actual project, the App Host should come first. Apologies if I didn't clarify that in the discussion. If someone, say, adds a Redis component to their front-end web project before having the call to `AddRedis` in the App Host, things will break when they F5, right? I'm optimizing for the F5 experience.

@bradygaster - I think this is optimized for the <kbd>F5</kbd>. If you can give me a list of which articles should have the order you specified, I can update them, but for now I'm moving forward with what I have in this PR.

**Pull request merged**